File: operationsresearchbasics.py
from scipy.optimize import linprog
import networkx as nx
import numpy as np

# 1. Linear Programming
def linear_programming_example():
    print("1. Linear Programming")
    # Maximize: 3x + 2y
    # Subject to:
    # x + y <= 4
    # 2x + y <= 5
    # x, y >= 0
    c = [-3, -2]  # Coefficients for the objective function (negated for maximization)
    A = [[1, 1], [2, 1]]  # Coefficients for inequality constraints
    b = [4, 5]  # Right-hand side of the inequalities
    bounds = [(0, None), (0, None)]  # x, y >= 0
    
    result = linprog(c, A_ub=A, b_ub=b, bounds=bounds, method='highs')
    print("Optimal value:", -result.fun)
    print("Optimal solution:", result.x)
    print()

# 2. Network Optimization (Shortest Path)
def network_optimization_example():
    print("2. Network Optimization (Shortest Path)")
    G = nx.DiGraph()
    G.add_weighted_edges_from([
        ('A', 'B', 4),
        ('A', 'C', 2),
        ('B', 'C', 1),
        ('B', 'D', 5),
        ('C', 'D', 8),
        ('C', 'E', 10),
        ('D', 'E', 2),
        ('D', 'Z', 6),
        ('E', 'Z', 3)
    ])
    shortest_path = nx.shortest_path(G, source='A', target='Z', weight='weight')
    shortest_distance = nx.shortest_path_length(G, source='A', target='Z', weight='weight')
    print("Shortest path from A to Z:", shortest_path)
    print("Shortest distance:", shortest_distance)
    print()

# 3. Transportation Problem
def transportation_problem_example():
    print("3. Transportation Problem")
    # Cost matrix (rows: sources, columns: destinations)
    costs = np.array([
        [2, 3, 1],
        [5, 4, 8],
        [5, 9, 2]
    ])
    supply = [20, 30, 25]  # Supply available at each source
    demand = [30, 25, 20]  # Demand required at each destination
    
    # Solve using linear programming
    num_sources = len(supply)
    num_destinations = len(demand)
    
    c = costs.flatten()  # Flatten cost matrix to vector
    A_eq = []
    
    # Supply constraints
    for i in range(num_sources):
        row = [1 if j // num_destinations == i else 0 for j in range(num_sources * num_destinations)]
        A_eq.append(row)
    
    # Demand constraints
    for k in range(num_destinations):
        row = [1 if j % num_destinations == k else 0 for j in range(num_sources * num_destinations)]
        A_eq.append(row)
    
    b_eq = supply + demand
    bounds = [(0, None) for _ in range(num_sources * num_destinations)]
    
    result = linprog(c, A_eq=np.array(A_eq), b_eq=b_eq, bounds=bounds, method='highs')
    print("Optimal cost:", result.fun)
    print("Optimal flows:")
    print(result.x.reshape(num_sources, num_destinations))
    print()

# 4. Simulation (Queueing System)
def queue_simulation_example():
    print("4. Simulation (Queueing System)")
    import random
    random.seed(42)
    
    arrival_rate = 5  # Customers per hour
    service_rate = 6  # Customers per hour
    num_customers = 10
    
    interarrival_times = [random.expovariate(arrival_rate) for _ in range(num_customers)]
    service_times = [random.expovariate(service_rate) for _ in range(num_customers)]
    
    arrival_times = np.cumsum(interarrival_times)
    start_times = np.zeros(num_customers)
    departure_times = np.zeros(num_customers)
    
    for i in range(num_customers):
        if i == 0:
            start_times[i] = arrival_times[i]
        else:
            start_times[i] = max(arrival_times[i], departure_times[i - 1])
        departure_times[i] = start_times[i] + service_times[i]
    
    print("Arrival times:", arrival_times)
    print("Service start times:", start_times)
    print("Departure times:", departure_times)
    print("Total time in system:", departure_times - arrival_times)
    print()

# Run examples
if __name__ == "__main__":
    linear_programming_example()
    network_optimization_example()
    transportation_problem_example()
    queue_simulation_example()

File: operationsresearchbasics.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <limits>
#include <algorithm>
#include <cmath>
#include <iomanip>
#include <cstdlib>
#include <ctime>

using namespace std;

// 1. Linear Programming using Simplex Method (Example)
void linearProgrammingExample() {
    cout << "1. Linear Programming" << endl;
    // Maximize: 3x + 2y
    // Subject to:
    // x + y <= 4
    // 2x + y <= 5
    // x, y >= 0

    // Coefficients of the objective function
    double c[] = {3, 2};

    // Coefficients of the constraints
    double A[2][2] = {{1, 1}, {2, 1}};
    double b[] = {4, 5};

    // Using brute force as a simple approximation
    double max_value = -numeric_limits<double>::infinity();
    pair<double, double> solution = {0, 0};

    for (double x = 0; x <= 4; x += 0.1) {
        for (double y = 0; y <= 4; y += 0.1) {
            if (x + y <= 4 && 2 * x + y <= 5) {
                double value = 3 * x + 2 * y;
                if (value > max_value) {
                    max_value = value;
                    solution = {x, y};
                }
            }
        }
    }

    cout << "Optimal value: " << max_value << endl;
    cout << "Optimal solution: x = " << solution.first << ", y = " << solution.second << endl;
    cout << endl;
}

// 2. Network Optimization (Shortest Path using Dijkstra's Algorithm)
void networkOptimizationExample() {
    cout << "2. Network Optimization (Shortest Path)" << endl;

    const int INF = numeric_limits<int>::max();
    int n = 6; // Number of nodes
    vector<vector<pair<int, int>>> graph(n);

    // Edges: (source, destination, weight)
    vector<tuple<int, int, int>> edges = {
        {0, 1, 4}, {0, 2, 2}, {1, 2, 1}, {1, 3, 5},
        {2, 3, 8}, {2, 4, 10}, {3, 4, 2}, {3, 5, 6}, {4, 5, 3}};

    for (auto &[u, v, w] : edges) {
        graph[u].emplace_back(v, w);
    }

    int source = 0, target = 5;
    vector<int> dist(n, INF), parent(n, -1);
    dist[source] = 0;

    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;
    pq.emplace(0, source);

    while (!pq.empty()) {
        auto [d, u] = pq.top();
        pq.pop();
        if (d > dist[u])
            continue;

        for (auto &[v, w] : graph[u]) {
            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
                parent[v] = u;
                pq.emplace(dist[v], v);
            }
        }
    }

    // Reconstruct shortest path
    vector<int> path;
    for (int v = target; v != -1; v = parent[v]) {
        path.push_back(v);
    }
    reverse(path.begin(), path.end());

    cout << "Shortest path: ";
    for (int node : path)
        cout << node << " ";
    cout << endl;
    cout << "Shortest distance: " << dist[target] << endl;
    cout << endl;
}

// 3. Simulation (Queueing System)
void queueSimulationExample() {
    cout << "3. Simulation (Queueing System)" << endl;

    srand(static_cast<unsigned>(time(0)));

    int num_customers = 10;
    double arrival_rate = 5;  // Customers per hour
    double service_rate = 6; // Customers per hour

    vector<double> interarrival_times(num_customers);
    vector<double> service_times(num_customers);
    vector<double> arrival_times(num_customers);
    vector<double> start_times(num_customers);
    vector<double> departure_times(num_customers);

    for (int i = 0; i < num_customers; ++i) {
        interarrival_times[i] = -log((double)rand() / RAND_MAX) / arrival_rate;
        service_times[i] = -log((double)rand() / RAND_MAX) / service_rate;
    }

    arrival_times[0] = interarrival_times[0];
    start_times[0] = arrival_times[0];
    departure_times[0] = start_times[0] + service_times[0];

    for (int i = 1; i < num_customers; ++i) {
        arrival_times[i] = arrival_times[i - 1] + interarrival_times[i];
        start_times[i] = max(arrival_times[i], departure_times[i - 1]);
        departure_times[i] = start_times[i] + service_times[i];
    }

    cout << "Customer | Arrival Time | Start Time | Departure Time | Total Time in System" << endl;
    for (int i = 0; i < num_customers; ++i) {
        cout << setw(8) << i + 1 << " | " << setw(12) << fixed << setprecision(2) << arrival_times[i]
             << " | " << setw(10) << start_times[i] << " | " << setw(14) << departure_times[i]
             << " | " << setw(20) << departure_times[i] - arrival_times[i] << endl;
    }
    cout << endl;
}

// Main function to run examples
int main() {
    linearProgrammingExample();
    networkOptimizationExample();
    queueSimulationExample();
    return 0;
}
